In [115]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import sklearn

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.linear_model import ElasticNet

In [116]:
dataframe = pd.read_csv("communities.csv")
dataframe

8   ?    ?.1         Lakewoodcity   1  0.19  0.33  0.02   0.9  0.12  \
0     53   ?      ?          Tukwilacity   1  0.00  0.16  0.12  0.74  0.45   
1     24   ?      ?         Aberdeentown   1  0.00  0.42  0.49  0.56  0.17   
2     34   5  81440  Willingborotownship   1  0.04  0.77  1.00  0.08  0.12   
3     42  95   6096    Bethlehemtownship   1  0.01  0.55  0.02  0.95  0.09   
4      6   ?      ?    SouthPasadenacity   1  0.02  0.28  0.06  0.54  1.00   
...   ..  ..    ...                  ...  ..   ...   ...   ...   ...   ...   
1988  12   ?      ?    TempleTerracecity  10  0.01  0.40  0.10  0.87  0.12   
1989   6   ?      ?          Seasidecity  10  0.05  0.96  0.46  0.28  0.83   
1990   9   9  80070        Waterburytown  10  0.16  0.37  0.25  0.69  0.04   
1991  25  17  72600          Walthamcity  10  0.08  0.51  0.06  0.87  0.22   
1992   6   ?      ?          Ontariocity  10  0.20  0.78  0.14  0.46  0.24   

      ...  0.12.2  0.26.1  0.2.1  0.06.3  0.04.2  0.9.1  0.5.2  0.32.2  \
0     ...    0.02    0.12   0.45       ?       ?      ?      ?    0.00   
1     ...    0.01    0.21   0.02       ?       ?      ?      ?    0.00   
2     ...    0.02    0.39   0.28       ?       ?      ?      ?    0.00   
3     ...    0.04    0.09   0.02       ?       ?      ?      ?    0.00   
4     ...    0.01    0.58   0.10       ?       ?      ?      ?    0.00   
...   ...     ...     ...    ...     ...     ...    ...    ...     ...   
1988  ...    0.01    0.28   0.05       ?       ?      ?      ?    0.00   
1989  ...    0.02    0.37   0.20       ?       ?      ?      ?    0.00   
1990  ...    0.08    0.32   0.18    0.08    0.06   0.78      0    0.91   
1991  ...    0.03    0.38   0.33    0.02    0.02   0.79      0    0.22   
1992  ...    0.11    0.30   0.05    0.08    0.04   0.73    0.5    1.00   

      0.14.3  0.2.2  
0          ?   0.67  
1          ?   0.43  
2          ?   0.12  
3          ?   0.03  
4          ?   0.14  
...      ...    ...  
1988       ?   0.09  
1989       ?   0.45  
1990    0.28   0.23  
1991    0.18   0.19  
1992    0.13   0.48  

[1993 rows x 128 columns]

In [117]:
df = dataframe.replace("?", np.nan)
df = df.dropna(axis=1)
df = df.drop(["Lakewoodcity"], axis=1)
df

8   1  0.19  0.33  0.02   0.9  0.12  0.17  0.34  0.47  ...  0.12.1  \
0     53   1  0.00  0.16  0.12  0.74  0.45  0.07  0.26  0.59  ...    0.21   
1     24   1  0.00  0.42  0.49  0.56  0.17  0.04  0.39  0.47  ...    0.14   
2     34   1  0.04  0.77  1.00  0.08  0.12  0.10  0.51  0.50  ...    0.19   
3     42   1  0.01  0.55  0.02  0.95  0.09  0.05  0.38  0.38  ...    0.11   
4      6   1  0.02  0.28  0.06  0.54  1.00  0.25  0.31  0.48  ...    0.70   
...   ..  ..   ...   ...   ...   ...   ...   ...   ...   ...  ...     ...   
1988  12  10  0.01  0.40  0.10  0.87  0.12  0.16  0.43  0.51  ...    0.22   
1989   6  10  0.05  0.96  0.46  0.28  0.83  0.32  0.69  0.86  ...    0.53   
1990   9  10  0.16  0.37  0.25  0.69  0.04  0.25  0.35  0.50  ...    0.25   
1991  25  10  0.08  0.51  0.06  0.87  0.22  0.10  0.58  0.74  ...    0.45   
1992   6  10  0.20  0.78  0.14  0.46  0.24  0.77  0.50  0.62  ...    0.68   

      0.42  0.5.1  0.51.1  0.64  0.12.2  0.26.1  0.2.1  0.32.2  0.2.2  
0     0.50   0.34    0.60  0.52    0.02    0.12   0.45    0.00   0.67  
1     0.49   0.54    0.67  0.56    0.01    0.21   0.02    0.00   0.43  
2     0.30   0.73    0.64  0.65    0.02    0.39   0.28    0.00   0.12  
3     0.72   0.64    0.61  0.53    0.04    0.09   0.02    0.00   0.03  
4     0.42   0.49    0.73  0.64    0.01    0.58   0.10    0.00   0.14  
...    ...    ...     ...   ...     ...     ...    ...     ...    ...  
1988  0.28   0.34    0.48  0.39    0.01    0.28   0.05    0.00   0.09  
1989  0.25   0.17    0.10  0.00    0.02    0.37   0.20    0.00   0.45  
1990  0.68   0.61    0.79  0.76    0.08    0.32   0.18    0.91   0.23  
1991  0.64   0.54    0.59  0.52    0.03    0.38   0.33    0.22   0.19  
1992  0.50   0.34    0.35  0.68    0.11    0.30   0.05    1.00   0.48  

[1993 rows x 102 columns]

In [118]:
correlation = df.corr()
correlation

8         1      0.19      0.33      0.02       0.9      0.12  \
8       1.000000 -0.003851 -0.078962 -0.050049 -0.046576  0.214475 -0.318834   
1      -0.003851  1.000000 -0.006739  0.034876  0.003032 -0.003968 -0.001445   
0.19   -0.078962 -0.006739  1.000000 -0.045742  0.231594 -0.301269  0.181738   
0.33   -0.050049  0.034876 -0.045742  1.000000 -0.067385 -0.235722  0.201965   
0.02   -0.046576  0.003032  0.231594 -0.067385  1.000000 -0.794350 -0.106800   
...          ...       ...       ...       ...       ...       ...       ...   
0.12.2 -0.040052  0.018251  0.713630 -0.014877  0.149941 -0.131560 -0.001044   
0.26.1 -0.137911 -0.012473  0.231892 -0.004018  0.095105 -0.337530  0.389959   
0.2.1  -0.021334 -0.000410  0.270344 -0.051446  0.147092 -0.215707  0.296938   
0.32.2 -0.074756 -0.023061  0.466091 -0.094019  0.261174 -0.276603  0.101987   
0.2.2  -0.211587 -0.032725  0.367346 -0.034995  0.631279 -0.684787  0.037609   

            0.17      0.34      0.47  ...    0.12.1      0.42     0.5.1  \
8      -0.210856  0.100326  0.036860  ... -0.356569  0.340235  0.151794   
1       0.001003 -0.000578  0.006954  ... -0.000520 -0.026151 -0.022121   
0.19    0.156202  0.006654  0.130468  ...  0.173003 -0.124710 -0.123624   
0.33    0.468785  0.520365  0.367334  ...  0.311003 -0.049650 -0.057931   
0.02   -0.066552  0.122188  0.153438  ... -0.098258  0.093231 -0.045712   
...          ...       ...       ...  ...       ...       ...       ...   
0.12.2  0.011607  0.025875  0.062522  ... -0.048747 -0.069060 -0.116861   
0.26.1  0.370099 -0.095720  0.119737  ...  0.614829 -0.221328  0.020715   
0.2.1   0.078357 -0.173248 -0.023493  ...  0.401766 -0.175421  0.192386   
0.32.2  0.125329  0.001558  0.117396  ...  0.171186 -0.133852 -0.090179   
0.2.2   0.293062  0.060438  0.153345  ...  0.194366 -0.077252 -0.155406   

          0.51.1      0.64    0.12.2    0.26.1     0.2.1    0.32.2     0.2.2  
8       0.061749  0.127630 -0.040052 -0.137911 -0.021334 -0.074756 -0.211587  
1      -0.035683 -0.028000  0.018251 -0.012473 -0.000410 -0.023061 -0.032725  
0.19    0.001914 -0.089632  0.713630  0.231892  0.270344  0.466091  0.367346  
0.33   -0.103373 -0.003301 -0.014877 -0.004018 -0.051446 -0.094019 -0.034995  
0.02    0.056185 -0.002913  0.149941  0.095105  0.147092  0.261174  0.631279  
...          ...       ...       ...       ...       ...       ...       ...  
0.12.2 -0.062853 -0.104396  1.000000 -0.170901  0.007658  0.299481  0.196855  
0.26.1  0.118132 -0.012162 -0.170901  1.000000  0.587800  0.228073  0.281404  
0.2.1   0.081317 -0.042420  0.007658  0.587800  1.000000  0.206612  0.153845  
0.32.2  0.008504 -0.107286  0.299481  0.228073  0.206612  1.000000  0.348784  
0.2.2   0.075517 -0.019464  0.196855  0.281404  0.153845  0.348784  1.000000  

[102 rows x 102 columns]

In [119]:
X = df.drop("0.2.2", axis=1)
y = df["0.2.2"]

In [120]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [121]:
LinReg = LinearRegression()
LinReg.fit(X_train, y_train)

LinearRegression()

In [122]:
print(f"train_score: {LinReg.score(X_train, y_train)}")
print(f"test_score: {LinReg.score(X_test, y_test)}")

train_score: 0.7004538296852882
test_score: 0.6563082285965177


In [123]:
y_pred = LinReg.predict(X_test)
compare = pd.DataFrame({"y_test" : y_test, "y_pred" : y_pred})
compare

y_test    y_pred
889     0.20  0.042846
1672    0.16  0.109728
414     0.14  0.133445
1599    0.04  0.009293
849     0.06  0.211507
...      ...       ...
909     0.04 -0.058810
261     0.67  0.721433
576     0.04  0.035972
963     0.01  0.049506
240     0.12  0.223519

[399 rows x 2 columns]

In [124]:
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
print(f'Mean absolute error: {mae:.2f}')
print(f'Mean squared error: {mse:.2f}')
print(f'Root mean squared error: {rmse:.2f}')

Mean absolute error: 0.09
Mean squared error: 0.02
Root mean squared error: 0.14


---

**L1**

In [125]:
lasso1 = Lasso(alpha=1.0).fit(X_train, y_train)
print(f"Score for train set {lasso1.score(X_train, y_train)}")
print(f"Score for test set {lasso1.score(X_test, y_test)}")

Score for train set 0.0
Score for test set -0.0006155179086768658


In [126]:
lasso2 = Lasso(alpha=0.01).fit(X_train, y_train)
print(f"Score for train set {lasso2.score(X_train, y_train)}")
print(f"Score for test set {lasso2.score(X_test, y_test)}")

Score for train set 0.5704543880704758
Score for test set 0.5669282571097347


**L2**

In [127]:
ridge = Ridge(alpha=0.7).fit(X_train, y_train)
print(f"Score for train set {ridge.score(X_train, y_train)}")
print(f"Score for test set {ridge.score(X_test, y_test)}")

Score for train set 0.695643908991169
Score for test set 0.6632185752127464


In [128]:
ridge = Ridge(alpha=0.9).fit(X_train, y_train)
print(f"Score for train set {ridge.score(X_train, y_train)}")
print(f"Score for test set {ridge.score(X_test, y_test)}")

Score for train set 0.6946808793583663
Score for test set 0.6633398850763584


**L1 + L2**

In [129]:
eklastic = ElasticNet(alpha=0.01, l1_ratio=0.01).fit(X_train, y_train)
print(f"Score for train set {eklastic.score(X_train, y_train)}")
print(f"Score for test set {eklastic.score(X_test, y_test)}")

Score for train set 0.6711994262028338
Score for test set 0.6560692256706928


**None of the Regularization methodes exceeded our model train score, but L2 made a micro better test score. Deviation at 0.14 after the test of our model might be indicating this.**